### PR TITLE
[SYCLomatic] Support 30 cusolver API migration

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/lapack_utils.hpp.inc [hunk 1]
```diff
@@ -834,13 +837,311 @@ template <typename T> struct potrs_impl {
...
+    working_memory<std::int64_t> m_device(1, q);
+    auto z_data = z.get_memory();
+    auto m_device_data = m_device.get_memory();
+    auto a_data = dpct::detail::get_memory(reinterpret_cast<T *>(a));
+    auto b_data = dpct::detail::get_memory(reinterpret_cast<T *>(b));
+    auto device_ws_data =
+        dpct::detail::get_memory(reinterpret_cast<T *>(device_ws));
+    auto vl_value =
+        *reinterpret_cast<typename value_type_trait<T>::value_type *>(vl);
+    auto vu_value =
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/lapack_utils.hpp.inc]: using value_type = typename value_type_trait<T>::value_type

File: clang/runtime/dpct-rt/include/lapack_utils.hpp.inc [hunk 2]
```diff
@@ -1315,6 +1623,413 @@ inline int potrs(sycl::queue &q, oneapi::mkl::uplo uplo, std::int64_t n,
...
+                                   oneapi::mkl::rangev range,
+                                   oneapi::mkl::uplo uplo, std::int64_t n,
+                                   library_data_t a_type, std::int64_t lda,
+                                   void *vl, void *vu, std::int64_t il,
+                                   std::int64_t iu, library_data_t w_type,
+                                   std::size_t *device_ws_size,
+                                   std::size_t *host_ws_size = nullptr) {
+  if (host_ws_size)
+    *host_ws_size = 0;
+  oneapi::mkl::compz compz_jobz;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/lapack_utils.hpp.inc]: move line 1667 to 1673 to detail namespace

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp [hunk 3]
```diff
@@ -665,6 +668,240 @@ template <typename T> struct potrs_impl {
...
+#endif
+  }
+};
+
+template <typename T> struct syhegvx_scratchpad_size_impl {
+  void operator()(sycl::queue &q, std::int64_t itype, oneapi::mkl::compz jobz,
+                  oneapi::mkl::rangev range, oneapi::mkl::uplo uplo,
+                  std::int64_t n, std::int64_t lda, std::int64_t ldb, void *vl,
+                  void *vu, std::int64_t il, std::int64_t iu,
+                  std::size_t *device_ws_size) {
```
Comment (Alexander-Kleymenov) [inline clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: Just noting that passing `device_ws_size` by reference may make this a safer call.

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp [hunk 4]
```diff
@@ -665,6 +668,240 @@ template <typename T> struct potrs_impl {
...
+                  std::int64_t lda, void *b, std::int64_t ldb, void *w,
+                  void *device_ws, std::size_t device_ws_size,
+                  int *info) {
+    auto a_data = dpct::detail::get_memory(reinterpret_cast<T *>(a));
+    auto b_data = dpct::detail::get_memory(reinterpret_cast<T *>(b));
+    auto device_ws_data =
+        dpct::detail::get_memory(reinterpret_cast<T *>(device_ws));
+    if constexpr (std::is_same_v<T, float> || std::is_same_v<T, double>) {
+      auto w_data = dpct::detail::get_memory(reinterpret_cast<T *>(w));
+      oneapi::mkl::lapack::sygvd(q, itype, jobz, uplo, n, a_data, lda, b_data,
```
Comment (Alexander-Kleymenov) [inline clang/test/dpct/helper_files_ref/include/lapack_utils.hpp:891]: Looks good assuming this is an ordered queue and no event dependencies have to be handled..

**Review by Alexander-Kleymenov — COMMENTED** [review]
Looks good!

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp (hunk 3 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: fixed

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/lapack_utils.hpp:891]: Yes. It must be an ordered queue. I have updated the comments in the code.

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp [hunk 5]
```diff
@@ -97,7 +98,8 @@ inline int sygvd(sycl::queue &queue, std::int64_t itype, oneapi::mkl::job jobz,
 /// generalized Hermitian positive-definite eigenproblem using a divide and
 /// conquer method.
 /// \return Returns 0 if no synchronous exception, otherwise returns 1.
-/// \param [in] queue Device queue where calculations will be performed.
+/// \param [in] queue Device queue where calculations will be performed. It must
+/// have the in_order property when using the USM mode.
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: ### 
```suggestion
/// have the in_order property when using the USM mode (DPCT_USM_LEVEL_NONE is not defined).
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**